core/kernel/tee_ta_manager.c

```diff
@@ -261,6 +262,93 @@ static void tee_ta_unlink_session(struct tee_ta_session *s,
 	mutex_unlock(&tee_ta_mutex);
 }
 
+static void destroy_session(struct tee_ta_session *s,
+			    struct tee_ta_session_head *open_sessions)
+{
+	tee_ta_unlink_session(s, open_sessions);
+#if defined(CFG_TA_GPROF_SUPPORT)
+	free(s->sbuf);
+#endif
+	free(s);
+}
+
+static void destroy_context(struct tee_ta_ctx *ctx)
+{
+	DMSG("Destroy TA ctx (0x%" PRIxVA ")",  (vaddr_t)ctx);
+
+	condvar_destroy(&ctx->busy_cv);
+	pgt_flush_ctx(ctx);
+	ctx->ops->destroy(ctx);
+}
+
+static void destroy_ta_ctx_from_session(struct tee_ta_session *s)
+{
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	struct user_ta_ctx *utc = NULL;
+	size_t count = 1; /* start counting the references to the context */
+
+	DMSG("Remove references to context (0x%" PRIxVA ")", (vaddr_t)s->ctx);
+
+	mutex_lock(&tee_ta_mutex);
+	nsec_sessions_list_head(&open_sessions);
+
+	/*
+	 * Next two loops will remove all references to the context which is
+	 * about to be destroyed, but avoiding such operation to the current
+	 * session. That will be done later in this function, only after
+	 * the context will be properly destroyed.
+	 */
+
+	/*
+	 * Scan the entire list of opened sessions by the clients from
+	 * non-secure world.
+	 */
+	TAILQ_FOREACH(sess, open_sessions, link) {
+		if (sess->ctx == s->ctx && sess != s) {
+			sess->ctx = NULL;
+			count++;
+		}
+	}
+
+	/*
+	 * Scan all sessions opened from secure side by searching through
+	 * all available TA instances and for each context, scan all opened
+	 * sessions.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+		if (is_user_ta_ctx(ctx)) {
+			utc = to_user_ta_ctx(ctx);
+
+			TAILQ_FOREACH(sess, &utc->open_sessions, link) {
+				if (sess->ctx == s->ctx && sess != s) {
+					sess->ctx = NULL;
+					count++;
+				}
+			}
+		}
+	}
+
+	assert(count == s->ctx->ref_count);
+
+	TAILQ_REMOVE(&tee_ctxes, s->ctx, link);
+	mutex_unlock(&tee_ta_mutex);
+
+	/*
+	 * Once there are no more references to the context of
+	 * TA panicked instance, release the allocated memory and
+	 * destroy the object.
+	 */
+	destroy_context(s->ctx);
+
+	/*
+	 * Mark the panicked context in order to make it
+	 * unusable in the future.
+	 */
```

Comment:
You can remove this comment, I guess it's obvious as already done for other context references in the 2 loops above.